# Introduction - Hashtables

In [1]:
import numpy as np
n = 5_000_000
data = list(np.random.randint(low=0,high=1_000_000,size=n))
data[0:10]

[13776, 897305, 15609, 948002, 970629, 37020, 70420, 719094, 322424, 362007]

## Linear Search

In [2]:
def linear_search(data, x):
    for item in data:
        if item == x:
            return True
        
    return False

### Search for one integer

In [3]:
x = %time linear_search(data, 999)

CPU times: user 39.9 ms, sys: 1.53 ms, total: 41.4 ms
Wall time: 40.1 ms


### Search for multiple integers

In [4]:
%time for x in range(50): linear_search(data, x)

CPU times: user 5.68 s, sys: 25.6 ms, total: 5.71 s
Wall time: 5.71 s


### Hashtables and Hashsearch Implementation

In [5]:
def hash_code(x):
    return x % 10  # put x into one of 10 possible buckets


def hash_table(data):
    buckets = [[] for i in range(10)]  # create 10 empty buckets
    for d in data:
        buckets[hash_code(d)].append(d)  # partition the integers into 10 buckets based on its hashcode
    return buckets


def hash_search(buckets, x):
    i = hash_code(x)
    for a in buckets[i]:
        if a == x:
            return True
    return False

### STEP 1: create buckets from the data

In [6]:
buckets = hash_table(data)  # STEP 1: create buckets from the data

### STEP 2: perform hash search for one integer

In [7]:
%time hash_search(buckets, 999)  # STEP 2: perform hash search for one integer

CPU times: user 1.88 ms, sys: 1 µs, total: 1.88 ms
Wall time: 1.88 ms


True

### STEP 3: perform hash search for multiple integers

In [8]:
%time for x in range(50): hash_search(buckets, x) # STEP 3: perform hash search for multiple integers

CPU times: user 573 ms, sys: 8.24 ms, total: 581 ms
Wall time: 580 ms


#### Multiple Integers Search Time:
#### Linear Search: 5.71 s
#### After Implementing Hashtables: 580 ms